src/cockpit/router.py

```diff
@@ -26,12 +27,47 @@
 logger = logging.getLogger(__name__)
 
 
+class ExecutionQueue(collections.deque):
+    def run(self):
+        for method, args in self:
+            method(*args)
+
+    def wrap(self, method):
+        """Temporarily delay calls to a given class method.
+
+        Functions by replacing the named function at the instance __dict__
+        level, effectively providing an override for exactly one instance.
+        Queues the invocations.  Run them later with .run(). Reverse this call
+        by deleting the named methods from the instance.
+        """
+        # NB: this function is stored in the instance dict and therefore
+        # doesn't function as a descriptor, isn't a method, doesn't get bound,
+        # and therefore doesn't receive a self parameter
+        setattr(method.__self__, method.__func__.__name__, lambda *args: self.append((method, args)))
+
+
 class Endpoint:
     router: Router
+    queue: Optional[ExecutionQueue] = None
 
     def __init__(self, router: Router):
         self.router = router
 
+    def freeze_endpoint(self):
+        self.queue = ExecutionQueue()
+        self.queue.wrap(self.do_channel_control)
+        self.queue.wrap(self.do_channel_data)
+
+    def thaw_endpoint(self):
+        queue = self.queue
+
+        if queue is not None:
```

Comment:
Why the reassignment dance?

Reply:
Because we need to call `queue.run()` after setting `self.queue = None`.

Reply:
Right, but why do we have to?

Reply:
This works fine, and is a little more straightforward:
```diff
--- src/cockpit/router.py
+++ src/cockpit/router.py
@@ -69,11 +69,9 @@ class Endpoint:
         self.queue = ExecutionQueue({self.do_channel_control, self.do_channel_data})
 
     def thaw_endpoint(self):
-        queue = self.queue
-
-        if queue is not None:
+        if self.queue is not None:
+            self.queue.run()
             self.queue = None
-            queue.run()
 
     # interface for receiving messages
     def do_channel_control(self, channel: str, command: str, message: Dict[str, object]) -> None:
```